okhttp/src/main/java/com/squareup/okhttp/ConnectionSpec.java

```diff
@@ -143,35 +167,65 @@ void apply(SSLSocket sslSocket, Route route) {
         cipherSuitesToEnable = newEnabledCipherSuites;
       }
     }
-    // null means "use default set".
-    if (cipherSuitesToEnable != null) {
-      sslSocket.setEnabledCipherSuites(cipherSuitesToEnable);
-    }
 
-    Platform platform = Platform.get();
-    if (specToApply.supportsTlsExtensions) {
-      platform.configureTlsExtensions(sslSocket, route.address.uriHost, route.address.protocols);
-    }
+    String[] protocolsToSelectFrom = sslSocket.getEnabledProtocols();
+    String[] protocolsToEnable = Util.intersect(String.class, tlsVersions, protocolsToSelectFrom);
+    return new Builder(this)
+        .cipherSuites(cipherSuitesToEnable)
+        .tlsVersions(protocolsToEnable)
+        .build();
   }
 
   /**
-   * Returns a copy of this that omits cipher suites and TLS versions not
-   * enabled by {@code sslSocket}.
+   * Returns {@code true} if the socket, as currently configured, supports this ConnectionSpec.
+   * In order for a socket to be compatible the enabled cipher suites and protocols must match.
+   *
+   * <p>For cipher suites, at least one of the {@link #cipherSuites() required cipher suites} must
+   * match the socket's enabled cipher suites. If there are no required cipher suites the socket
+   * must have at least one cipher suite enable.
+   *
+   * <p>For protocols, the first {@link #tlsVersions() TLS version} is considered a minimum
+   * requirement and must be be present in the socket's enabled protocols. For example, a
+   * connection spec containing [TLSv1.2, TLSv1.1, etc.] is only compatible if the socket has
+   * TLSv1.2 enabled.
    */
-  private ConnectionSpec supportedSpec(SSLSocket sslSocket) {
-    String[] cipherSuitesToEnable = null;
-    if (cipherSuites != null) {
-      String[] cipherSuitesToSelectFrom = sslSocket.getEnabledCipherSuites();
-      cipherSuitesToEnable =
-          Util.intersect(String.class, cipherSuites, cipherSuitesToSelectFrom);
+  public boolean isCompatible(SSLSocket socket) {
```

Comment:
This method is used in two places:
1) To judge whether a socket is compatible with a given ConnectionSpec
2) To judge whether a socket is _going to be_ compatible with a future ConnectionSpec (in order to determine whether we need to create another plain socket).

In the first case we burn the plain socket in the attempt (if it fails), in the second we open sockets we might not be able to use.

It seemed better to trade CPU for I/O.

In this change (with the fallback modification) we also have more connection specs, and the isCompatible() is important to skip over those that don't apply to satisfy the "socket must support the primary protocol" logic.

There was a obvious optimization I could make here (in #1388) but it was mostly "terminate early on success". To optimize further I could see us creating a bitmask fingerprint of ciphers / protocols and memoizing it rather than keep doing all the string comparisons (may not work for ciphers if we need to deal with ciphers that we don't know about).